src/PluginStructs/P200_data_struct.cpp

```diff
@@ -0,0 +1,189 @@
+#include "../PluginStructs/P200_data_struct.h"
+
+
+// Needed also here for PlatformIO's library finder as the .h file 
+// is in a directory which is excluded in the src_filter
+#include <ESPeasySerial.h>
+#include <base64.h>
+
+#ifdef USES_P200
+
+
+P200_data_struct::P200_data_struct() :  easySerial(nullptr) {}
+
+P200_data_struct::~P200_data_struct() {
+  reset();
+}
+
+void P200_data_struct::reset() {
+  if (easySerial != nullptr) {
+    delete easySerial;
+    easySerial = nullptr;
+  }
+}
+
+bool P200_data_struct::init(ESPEasySerialPort port, const int16_t serial_rx, const int16_t serial_tx, unsigned long baudrate) {
+  if ((serial_rx < 0) && (serial_tx < 0)) {
+    return false;
+  }
+  reset();
+  easySerial = new (std::nothrow) ESPeasySerial(port, serial_rx, serial_tx);
+
+  if (isInitialized()) {
+    easySerial->begin(baudrate);
+    return true;
+  }
+  return false;
+}
+
+bool P200_data_struct::isInitialized() const {
+  return easySerial != nullptr;
+}
+
+bool P200_data_struct::loop() {
+  if (!isInitialized()) {
+    return false;
+  }
+  bool fullSentenceReceived = false;
+
+  if (easySerial != nullptr) {
+    int available = easySerial->available();
+
+    while (available > 0 && !fullSentenceReceived) {
+      char c = easySerial->read();
+      --available;
+
+      if (available == 0) {
+        available = easySerial->available();
+        if (available == 0) { resetSerialTimeout(); }
+        delay(0);
+      }
+
+      switch (sentence_part.length()) {
+        case 0:
+          if (c != '\r') break; // wait for '\r'
+        default:
+          sentence_part += c;
+          sentence_checksum -= c;
+          break;
+      }
+
+      if (max_length_reached()) { fullSentenceReceived = true; }
+    }
+  }
+
+  if (serialTimeout() || fullSentenceReceived) {
+    const size_t length = sentence_part.length();
+    bool valid          = length > 0;
+
+    fullSentenceReceived = false;
+
+    if (valid) {
+    // message or full buffer
+      
+      // RAW
+      if (output_type == 2) {
```

Comment:
Magic numbers should be avoided, better replace by `if (output_type == P200_OUTPUT_RAW) {` (move the `#define`'s from the .ino to .h file)

Reply:
Moved #defines to .h. Removed magic numbers.